### PR TITLE
(major/non-breaking?) Modules Refactor

### LINKED ISSUE #82: next (any) release to-do list
- [x] Update since references to TBA in `StickyAPI.java` (from #77)
- [x] Add since references to #92. (especially StringUtil changes) 

- [x]  (optional) Add since references to #72.

### LINKED ISSUE #94: Gradle module refactor
**Describe the solution you'd like**
Refactor StickyAPI into independent Gradle modules. These can be included in a plugin jar as required, rather than shading the entire API. It would also have benefits from a development perspective - less Git conflicts, easier code management etc.

Proposed modules to add:
- `bukkit` - Bukkit-related utilites, only works with the Craftbukkit runtime.
- `bungee` - Bungee-related utilites, only works with the Bungeecord runtime.
- `common` - Common utilites shared between the former, works without a specific runtime.
- `example` - An example module containing the required build files for people to copy - is not included in the final jar.
- `translation` - Module for dealing with string interpolations given configuration.
- `worlds` - Depending on its size, module for world management & templating.
- `math` - Math library and utilities.
- `config` - Library for better config management, with support for YAML, JSON, and TOML.

### PR DESCRIPTION
**Pull request by jcxldn**
## Notes

- CodeQL alerts will all be reset, since the source code paths have changed.
- Javadocs (both global ones and ones for each module) are now built during the build stage, so do we still need the javadoc-check ci job?

---------
## Partial(?) changelog from previous tagged release
---------
## StickyAPI v3 Changelog

**Upgrade Notes**
- Module support (dependency resolution change)
  - `com.dumbdogdiner:stickyapi` is still available, but is not recommended
    Please use `com.dumbdogdiner:stickyapi-(module)` instead.

- **breaking**: Config rewrite (#98)

**Internal (ci)**
- Added support for signed docs deployment
- Added GitHub CodeQL support

**Internal**
- Added private constructors to never instanciated classes

**Dependencies**
- `net.nemerosa.versioning`: `2.8.2` -> `2.14.0`
- `classgraph`: `4.8.98` -> `4.8.100`

**New Features**
- Build Info in `com.dumbdogdiner.stickyapi.StickyAPI` (#77)
- Added support for Stringified NBT Objects (#92)
- Added Book Text Utils (#72)
- NotNull annotations for `VoidGenerator`

**Deprecations**
- `getRandomNumber`
- `com.dumbdogdiner.stickyapi.bukkit.nms`

### CODE REVIEW

File: build.gradle
Comment (kaylendog) [inline build.gradle]: Worth changing this to something before merge.

File: build.gradle
Comment (kaylendog) [inline build.gradle:81]:
```groovy
compileJava {} // ?
```

Comment (kaylendog) [reply build.gradle:81]: Or does this break for the subproject block? Can't remember....

File: build.gradle
Comment (kaylendog) [inline build.gradle]: Consistent bracketing maybe?

File: build.gradle
Comment (kaylendog) [inline build.gradle:159]: Consistent `'`-ing

File: build.gradle
Comment (kaylendog) [inline build.gradle]: Same as above, does `javadoc {}` not exist?

File: bukkit/build.gradle
Comment (kaylendog) [inline bukkit/build.gradle]:
```suggestion
}

```

File: bukkit/src/main/java/com/dumbdogdiner/stickyapi/bukkit/nms/BukkitHandler.java
Comment (kaylendog) [inline bukkit/src/main/java/com/dumbdogdiner/stickyapi/bukkit/nms/BukkitHandler.java]:
```suggestion
 * @deprecated This will be superseded and removed in a future release.
```

File: bukkit/src/main/java/com/dumbdogdiner/stickyapi/bukkit/nms/PacketInjector.java
Comment (kaylendog) [inline bukkit/src/main/java/com/dumbdogdiner/stickyapi/bukkit/nms/PacketInjector.java]:
```suggestion
 * @deprecated This will be superseded and removed in a future release.
```

File: bukkit/src/main/java/com/dumbdogdiner/stickyapi/bukkit/nms/PacketInjector.java
Comment (kaylendog) [inline bukkit/src/main/java/com/dumbdogdiner/stickyapi/bukkit/nms/PacketInjector.java]:
```suggestion
     * @deprecated This will be superseded and removed in a future release.
```

File: bungee/build.gradle
Comment (kaylendog) [inline bungee/build.gradle]:
```suggestion
}

```

File: common/build.gradle
Comment (kaylendog) [inline common/build.gradle]:
```suggestion
tasks.compileJava.dependsOn licenseFormatMain, licenseFormatTest, copyMCFontExtractor

```

File: common/build.gradle
Comment (kaylendog) [inline common/build.gradle:70]: Would it not be more sensible if this was in the root project?

File: settings.gradle
Comment (kaylendog) [inline settings.gradle]:
```suggestion
include ":bungee"

```

**Review by kaylendog — COMMENTED** [review]
Some minor formatting complaints that I will fix in a second - pls see comments on the build files.

File: build.gradle
Comment (jcxldn) [reply build.gradle:81]: You can't use compilejava because of subprojects - it would only apply to the root project

File: build.gradle
Comment (jcxldn) [reply build.gradle]: See above comment

File: common/build.gradle
Comment (jcxldn) [reply common/build.gradle:70]: I've put it there for now as otherwise it's processing the entire project when it really only needs to process one file, and I'm not really sure about its behaviour in a root project either.

Comment (kaylendog) [reply common/build.gradle:70]: It's just odd that this prints information shared between the entire project...

Comment (jcxldn) [reply common/build.gradle:70]: Ah - the versioning plugin grabs information from git and gradle (our version + group is now set under allprojects {})

Comment (kaylendog) [reply common/build.gradle:70]: So... is this slightly pointless?

**Review by kaylendog — COMMENTED** [review]
Fixed formatting comments.

File: common/build.gradle
Comment (jcxldn) [reply common/build.gradle:70]: ? - This is from the PR for buildinfo and embeds version + commit details into StickyAPI.java at buildtime so plugins can expose exactly what version of StickyAPI they are using

Comment (kaylendog) [reply common/build.gradle:70]: Oh - can you leave a comment on this saying that's what this does? lul

**Review by jcxldn — COMMENTED** [review]
(ci-ops): 👍 [added local testing repo steps]

File: build.gradle
Comment (kaylendog) [inline build.gradle]: Modules should implement these on their own, should they not? This would give `bukkit` access to Bungeecord and vice versa....????? Ideally common shouldn't hard-depend on either and only use deps common between the two e.g. `ChatColor`

**Review by kaylendog — CHANGES_REQUESTED** [review]
Noticed weird dependencies in root build file...

File: build.gradle
Comment (jcxldn) [reply build.gradle]: ~~Blocked by #98.~~

Comment (kaylendog) [reply build.gradle]: Confusion?

File: bungee/src/main/java/com/dumbdogdiner/stickyapi/bungeecord/util/StartupUtil.java
Comment (kaylendog) [inline bungee/src/main/java/com/dumbdogdiner/stickyapi/bungeecord/util/StartupUtil.java:5]: This looks very similar to the Bukkit startup util - perhaps some de-dupe here?

File: common/serverversion/build.gradle
Comment (kaylendog) [inline common/serverversion/build.gradle]:
```suggestion
}

```

File: config/build.gradle
Comment (kaylendog) [inline config/build.gradle]:
```suggestion
}

```

File: config/src/main/java/com/dumbdogdiner/stickyapi/common/config/Configuration.java
Comment (kaylendog) [inline config/src/main/java/com/dumbdogdiner/stickyapi/common/config/Configuration.java]:
```suggestion
}

```

File: config/src/main/java/com/dumbdogdiner/stickyapi/common/config/providers/YamlProvider.java
Comment (kaylendog) [inline config/src/main/java/com/dumbdogdiner/stickyapi/common/config/providers/YamlProvider.java:104]: 4 tab indentation hhhhh - vlad will complain but idc >:3

File: settings.gradle
Comment (kaylendog) [inline settings.gradle]:
```suggestion
rootProject.name = "stickyapi"
```

**Review by kaylendog — COMMENTED** [review]
LGTM - some formatting thingies.

File: LICENSE_HEADER
Comment (kaylendog) [inline LICENSE_HEADER:4]:
```suggestion
 */
 
```

File: build.gradle
Comment (kaylendog) [inline build.gradle]: Do we want to include patch number?

File: common/build.gradle
Comment (kaylendog) [inline common/build.gradle]:
```suggestion
tasks.sources.finalizedBy commonSources

```

**Review by kaylendog — APPROVED** [review]
LGTM

**Pull request merged**
